nixos/postgresql: add freeformType to ensureClauses option

Mikilio opened the pull request:
This commit adds a freeform type to accomodate for new options to use with ALTER ROLE.
It also will perform extra checking and conversion for options with parameters.
Adds documented options for password and connection limit. Converts option names into upper case with space instead of underscore.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- Tested, as applicable:
  - [x] [NixOS tests] in [nixos/tests].
  - [x] [Package tests] at `passthru.tests`.
  - [x] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [x] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [x] Module addition: when adding a new NixOS module.
  - [x] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

@Ma27 @thoughtpolice @wolfgangwalther

Mikilio committed:
nixos/postgresql: add freeformType to ensureClauses option

This commit adds a freeform type to accomodate for new options to use
with ALTER ROLE.
It also will perform extra checking and conversion for options with
parameters.
Converts option names into upper case with space instead of underscore.

Ma27 reviewed (commented):
I think this setup is a mistake to begin with.

I'd be fine with allowing users to go the freeform-route, but I'd rather not maintain more option definitions tbh.

wolfgangwalther commented:
Well, the same level of validation that non-freeform-typing gives us. That means, checking whether the options passed are actually valid SQL. So I'd think a syntax check should be enough.

Mikilio commented:
@wolfgangwalther 
I added the validation.
I also updated the tests I locally tested with `nix build ".#nixosTests.postgresql.postgresql.postgresql_18.postgresql-clauses"`.

@Ma27
Should I keep or remove the documentation for the submodule options of ensureClauses?

Ma27 commented:
>  Should I keep or remove the documentation for the submodule options of ensureClauses?

Not a big fan ofd advertising this pattern in the first place, so I'm in favor of not adding extra options.

Mikilio commented:
Would you like to elaborate on what you recommend? I made this PR mostly because I was missing the ability to preconfigure a password. I saw the ensureClauses and it resembled the way I've seen that done many times in production. I decided to generalize the approach and make a PR upstream.

I am by no means a PostgreSQL guru, so I am happy to learn. I'd like to avoid adding technical debt or contributing to bad patterns.

wolfgangwalther commented:
The point of adding the *freeform* attribute is that you can configure any kind of option here - so it makes no sense to *explicitly* add `connection_limit` and `password`. So at the minimum, these additions should go away again.

But I'd go further, in a second commit, and remove the existing options, too - and make that thing *only* freeform.

Ma27 reviewed (commented):
Except for what we discussed, this change looks reasonable :+1:

Mikilio committed:
nixos/postgresql: remove explicit submodule options

These can be set via the freemform type anyway.

wolfgangwalther reviewed (commented):
Needs a rebase for a conflict about the test assertions, I believe.

Mikilio commented:
Also edited the Wiki https://wiki.nixos.org/w/index.php?title=PostgreSQL&oldid=prev&diff=29170

This is my first edit of the Wiki, so I am not familiar with the process.

wolfgangwalther reviewed (approved):
Made some minor changes, removing unrelated changes, moving a tiny bit from the second commit to the first commit, adjusted the commit messages a bit.

wolfgangwalther commented:
Thank you!
